processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -52,333 +55,395 @@
+        throw new ISE(
+            "start method returned false for lookup [%s]:[%s]",
+            lookupName,
+            lookupExtractorFactoryContainer
+        );
+      }
+
+      old = lookupMap.put(lookupName, lookupExtractorFactoryContainer);
+
+      LOG.debug("Loaded lookup [%s] with spec [%s].", lookupName, lookupExtractorFactoryContainer);
+
+      if (old != null) {
+        if (!old.getLookupExtractorFactory().close()) {
+          throw new ISE("close method returned false for lookup [%s]:[%s]", lookupName, old);
+        }
+      }
     }
-    finally {
-      startStopLock.readLock().unlock();
+
+    @Override
+    public String toString()
+    {
+      return "LoadNotice{" +
+             "lookupName='" + lookupName + '\'' +
+             ", lookupExtractorFactoryContainer=" + lookupExtractorFactoryContainer +
+             '}';
     }
   }
 
-  private void assertStarted() throws ISE
+  private class DropNotice implements Notice
   {
-    if (isClosed()) {
-      throw new ISE("lookup manager is closed");
+    private final String lookupName;
+
+    public DropNotice(String lookupName)
+    {
+      this.lookupName = lookupName;
     }
-  }
 
-  public boolean isClosed()
-  {
-    return !started.get();
+    @Override
+    public void handle(Map<String, LookupExtractorFactoryContainer> lookupMap)
+    {
+      final LookupExtractorFactoryContainer lookupExtractorFactoryContainer = lookupMap.remove(lookupName);
+
+      if (lookupExtractorFactoryContainer != null) {
+        LOG.debug("Removed lookup [%s] with spec [%s].", lookupName, lookupExtractorFactoryContainer);
+
+        if (!lookupExtractorFactoryContainer.getLookupExtractorFactory().close()) {
+          throw new ISE(
+              "close method returned false for lookup [%s]:[%s]",
+              lookupName,
+              lookupExtractorFactoryContainer
+          );
+        }
+      }
+    }
+
+    @Override
+    public String toString()
+    {
+      return "DropNotice{" +
+             "lookupName='" + lookupName + '\'' +
+             '}';
+    }
   }
 
-  private List<LookupBean> getAllAsList()
+  private class LookupUpdateState
   {
-    return Lists.newArrayList(
-        Collections2.transform(
-            lookupMap.entrySet(),
-            new Function<Map.Entry<String, LookupExtractorFactory>, LookupBean>()
-            {
-              @Nullable
-              @Override
-              public LookupBean apply(
-                  @Nullable
-                      Map.Entry<String, LookupExtractorFactory> input
-              )
-              {
-                final LookupBean lookupBean = new LookupBean();
-                lookupBean.factory = input.getValue();
-                lookupBean.name = input.getKey();
-                return lookupBean;
-              }
-            }
-        ));
+    private final ImmutableMap<String, LookupExtractorFactoryContainer> lookupMap;
+    private final ImmutableList<Notice> pendingNotices;
```

Comment:
Order could be preserved e. g. by using LinkedHashMap. Is it semantically meaningful not to skip requests? E. g. if someone sent load and then drop request, so that processing didn't take place in between